### PR TITLE
Gh 22 autodeploy to dockerhub

### LINKED ISSUE #22: Add code for releasing to Dockerhub
Add code into cd/deploy.sh for releasing to Dockerhub / updating version control from Travis CI when code is merged into master

When code is merged into master, Travis CI should:
* Tag the release in Git
* Push images to dockerhub
* Update the version on develop

### PR DESCRIPTION
**Pull request by d47853**
closes #22

### CODE REVIEW

**Comment by d47853** [comment]
Closing until wildfly changes are incorporated.

File: cd/deploy.sh [hunk 1]
```diff
@@ -1,5 +1,75 @@
...
+    app_version=$3
+
+    tags="$(echo "${version}" | sed -e "s/\(.*\)\.\(.*\)\..*/"${name}":"${version}"+"${app_version}" "${name}":"${version}" "${name}":\1.\2 "${name}":\1 "${name}":latest/")"
+
+    docker tag "${name}" "${tags}"
+    docker push "${tags}"
+}
+
+# If branch is not master or is pull request, exit
+if [ "${TRAVIS_PULL_REQUEST}" != "false"] || [ "${TRAVIS_BRANCH}" != "master" ]; then
```
Comment (ctas582) [inline cd/deploy.sh:23]:
```suggestion
if [ "${TRAVIS_PULL_REQUEST}" != "false" ] || [ "${TRAVIS_BRANCH}" != "master" ]; then
```

File: cd/deploy.sh [hunk 2]
```diff
@@ -1,5 +1,75 @@
 #!/bin/bash
 
-# Release to Dockerhub (if the branch is master)
+set -e
 
-# Update develop
\ No newline at end of file
+# Gets project root directory by calling two nested "dirname" commands on the this file 
+getRootDirectory() {
+    echo "$( cd "$(dirname "$(dirname "${BASH_SOURCE[0]}")")" > /dev/null 2>&1 && pwd )"
```
Comment (ctas582) [inline cd/deploy.sh:7]: I don't think this gets the root directory. It seems to get the directory that the script is being run from:

```
gaffer-docker$ ./cd/deploy.sh 
/home/ec2-user/gaffer-docker

cd$ ./deploy.sh 
/home/ec2-user/gaffer-docker/cd
```

File: cd/deploy.sh [hunk 3]
```diff
@@ -4,29 +4,27 @@ set -e
...
 # Pushes Tags to Dockerhub
 pushTags() {
     name=$1
     version=$2
     app_version=$3
-
-    tags="$(echo "${version}" | sed -e "s/\(.*\)\.\(.*\)\..*/"${name}":"${version}"+"${app_version}" "${name}":"${version}" "${name}":\1.\2 "${name}":\1 "${name}":latest/")"
-
+    tags="$(echo ${version} | sed -e "s|\(.*\)\.\(.*\)\..*|${name}:${version}+${app_version} ${name}:${version} ${name}:\1.\2 ${name}:\1 ${name}:latest|")"
     docker tag "${name}" "${tags}"
```
Comment (ctas582) [inline cd/deploy.sh:16]: I'm getting the error: `Error parsing reference: "gchq/hdfs:3.2.1+0.5.0 gchq/hdfs:3.2.1 gchq/hdfs:3.2 gchq/hdfs:3 gchq/hdfs:latest" is not a valid repository/tag: invalid reference format`. I don't think you can create multiple tags at once. Might have to loop through them, tagging and pushing each individually.

Comment (ctas582) [inline cd/deploy.sh:16]:
```suggestion
    docker tag "${name}:${version}" "${tags}"
```

**Comment by d47853** [comment]
Closing while fixes are applied

File: cd/deploy.sh [hunk 4]
```diff
@@ -12,9 +12,13 @@ pushTags() {
...
-    tags="$(echo ${version} | sed -e "s|\(.*\)\.\(.*\)\..*|${name}:${version}+${app_version} ${name}:${version} ${name}:\1.\2 ${name}:\1 ${name}:latest|")"
-    docker tag "${name}" "${tags}"
-    docker push "${tags}"
+    tags="$(echo ${version} | sed -e "s|\(.*\)\.\(.*\)\..*|${name}:${version}_build.${app_version} ${name}:${version} ${name}:\1.\2 ${name}:\1 ${name}:latest|")"
+    IFS=' '
+    read -a tagArray <<< "${tags}"
+    for tag in "${tagArray[@]}"; do
+        docker tag "${name}":"${version}" "${tag}"
+    done
+    docker push "${name}"
```
Comment (ctas582) [inline cd/deploy.sh:21]:
```suggestion
    for tag in "${tagArray[@]}"; do
        docker tag "${name}:${version}" "${tag}"
        docker push "${tag}"
    done
```

File: cd/update_app_version.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    old_version="$1"
+    major_and_minor="$(echo ${old_version} | sed 's|^\(.*\.\).*|\1|')"
+    bugfix="$(echo ${old_version} | sed 's|.*\.||')"
+    let "bugfix++"
+    echo "${major_and_minor}""${bugfix}"
+}
+
+# Performs a Find and replace on the Helm charts and the app_version file
+findAndReplace() {
+    find "$(getRootDirectory)" \( -iname Chart.y*ml -o -name app_version \) -exec sed -i'' -e "s:$1:$2:g" {} +
```
Comment (ctas582) [inline cd/update_app_version.sh:32]: Shouldn't be a problem most of the time, but if we ever have a repo/app version which matches the version of the ZooKeeper chart (currently 2.1.1) that we have as a dependency to the Gaffer chart, then this command will update that version number too.

File: cd/deploy.sh (hunk 4 above)
Comment (d47853) [reply cd/deploy.sh:21]: I would have thought pushing each tag one by one would be slower compared to pushing everything at the end?

File: cd/update_app_version.sh (hunk 5 above)
Comment (d47853) [reply cd/update_app_version.sh:32]: As you say, this shouldn't be a problem for now (as long as we're on a 0.x release and all dependencies are on a >0.x release). I'm not sure it's worth fixing now but worth raising an issue to fix in future.

File: cd/deploy.sh (hunk 4 above)
Comment (ctas582) [reply cd/deploy.sh]: `docker push "${name}"` would only push the 'latest' tag

File: cd/update_app_version.sh (hunk 5 above)
Comment (ctas582) [reply cd/update_app_version.sh:32]: 👍

**Pull request merged**